exec/testing/fake_exec.go

```diff
@@ -43,7 +46,23 @@ func (fake *FakeExec) Command(cmd string, args ...string) exec.Cmd {
 	}
 	i := fake.CommandCalls
 	fake.CommandCalls++
-	return fake.CommandScript[i](cmd, args...)
+	fakeCmd := fake.CommandScript[i](cmd, args...)
+	if fake.ExactOrder {
```

Comment:
Is there ever a time we don't want this?

Reply:
I can't really think of one, but I didn't want to suddenly break a bunch of people's tests by turning it on by default. As I am trying to have code in k/k use this package instead of the one found `k/k/pkg/util/mount`, I am already seeing that I am going to have to update a lot of tests because any calls to this fake exec without scripting those calls first results in the error: `panic: ran out of Command() actions. Could not handle command`

With this test module you are *required* to pre-load/script every call that's going to be made. Currently, it doesn't actually care what those calls are, just that there is the right number of them. Or you make less than what was scripted. :) Most of the code in k/k isn't doing that right now. I only found two places (linked below) that are using the `mount.Exec` callback to implement detailed command call verification. Most users just need a fake Exec that satisfies the API. So even without this change, there is going to be a lot of changes to tests to preload the calls that are made. Whether we should make it so that we also *always* check that those commands are called in the exact order given, with the exact arguments? Good question. I was leaving this feature off because it was a major change to how it works and seems like it broke backwards compat. I'll defer to advice from others. Thanks for review @thockin .

Examples of where no checking is done:
[1](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/fc/fc_test.go#L167) [2](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/testing/testing.go#L113)

And examples where it *is* done... 
[1](https://github.com/kubernetes/kubernetes/blob/faad5d52bcefa903a0adf4b1dea9a9413944ca33/pkg/util/mount/safe_format_and_mount_test.go#L193-L209) [2](https://github.com/kubernetes/kubernetes/blob/faad5d52bcefa903a0adf4b1dea9a9413944ca33/pkg/volume/iscsi/iscsi_util_test.go#L220-L240)

Note that in these examples, I'm trying to replace the code in k/k to use this library instead. The examples that do check cannot be moved without this equivalent feature without losing testing fidelity.

Reply:
A couple more thoughts on this...

Yes there are times where we won't want to validate arguments. Sometimes the arguments are not known ahead of time, because they reference tempdirs. One could make an argument for verifying the command, the command order, but maybe not the args? That would introduce a lot of knobs to turn if you did on a per-command basis.

I also think we need the ability to disable the current "script counting" -- meaning that this fake exec panics if you call a command and it hasn't already been scripted. I'm finding dozens of tests in k/k that just want a fake exec to satisfy an API and updating them all to script them ahead of time is going to be a massive amount of work.